### PR TITLE
Support parquet dereference pushdown for Iceberg Connector.

### PR DESCRIPTION
**Pull request by wypb**
```
== RELEASE NOTES ==

Iceberg Changes

* Add support for parquet dereference pushdown in Iceberg Connector.
```

image

### CODE REVIEW

**Comment by 397090770** [comment]
CC @agrawalreetika @ChunxuTang @rohanpednekar

**Review by wanglinsong — CHANGES_REQUESTED** [review]
https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#phrasing

**Comment by 397090770** [comment]
> https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines#phrasing Please use 'Add support ...' in release notes.

Thank you for your reply, I have modified it as required.

**Comment by wanglinsong** [comment]
> * Add support parquet dereference pushdown for Iceberg Connector.

how about

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java [hunk 1]
```diff
@@ -167,6 +168,7 @@ public void configure(Binder binder)
         configBinder(binder).bindConfig(OrcFileWriterConfig.class);
 
         binder.bind(IcebergPlanOptimizer.class).in(Scopes.SINGLETON);
+        binder.bind(IcebergParquetDereferencePushDown.class).in(Scopes.SINGLETON);
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java:171]: We probably don't need to bind the `IcebergParquetDereferencePushDown` here. Binding the plan optimizer looks sufficient.
Do you happen to have any specific reasons to bind it here?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java [hunk 2]
```diff
@@ -306,6 +323,35 @@ private static ConnectorPageSource createParquetPageSource(
...
+    {
+        switch (columnType) {
+            case REGULAR:
+                return REGULAR;
+            case SYNTHESIZED:
+                return SYNTHESIZED;
+        }
+
+        throw new PrestoException(GENERIC_INTERNAL_ERROR, "Unknown ColumnType: " + columnType);
+    }
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java:353]: This method looks a bit unclear to me. Why would you like to maintain a method to return REGULAR when REGULAR and SYNTHESIZED when SYNTHESIZED?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizerProvider.java [hunk 3]
```diff
@@ -25,19 +25,19 @@
 public class IcebergPlanOptimizerProvider
         implements ConnectorPlanOptimizerProvider
 {
-    private final ConnectorPlanOptimizer planOptimizer;
+    private final Set<ConnectorPlanOptimizer> planOptimizers;
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizerProvider.java:28]: Same as above. Any reasons to update the type to a set? Looks like this change is unrelated to the parquet deference pushdown.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergLogicalPlanner.java [hunk 4]
```diff
@@ -0,0 +1,437 @@
+/*
```
Comment (ChunxuTang) [inline presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergLogicalPlanner.java:1]: Quick question: Is this test based on the `TestHiveLogicalPlanner.java`?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergColumnHandle.java [hunk 5]
```diff
@@ -34,16 +42,27 @@
...
+    private final List<Subfield> requiredSubfields;
 
     @JsonCreator
     public IcebergColumnHandle(
             @JsonProperty("columnIdentity") ColumnIdentity columnIdentity,
             @JsonProperty("type") Type type,
-            @JsonProperty("comment") Optional<String> comment)
+            @JsonProperty("comment") Optional<String> comment,
+            @JsonProperty("columnType") ColumnType columnType,
+            @JsonProperty("requiredSubfields") List<Subfield> requiredSubfields)
```
[the remark this answers is not in the document]
Comment (397090770) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergColumnHandle.java:54]: The `requiredSubfields` field is required because it is used later to construct the ColumnIO object.

If we don't override the `withRequiredSubfields` method, we can maintain the `requiredSubfields` object without using a list.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java [hunk 6]
```diff
@@ -63,7 +63,7 @@
     private final List<PropertyMetadata<?>> tableProperties;
     private final ConnectorAccessControl accessControl;
     private final Set<Procedure> procedures;
-    private final ConnectorPlanOptimizer planOptimizer;
+    private final Set<ConnectorPlanOptimizer> planOptimizers;
```
[the remark this answers is not in the document]
Comment (397090770) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java:66]: Because `IcebergParquetDereferencePushDown` is implemented in a separate class, this causes the Iceberg connector to have two `ConnectorPlanOptimizer` implementation classes, the other is the `IcebergPlanOptimizer`. We need to pass these two` ConnectorPlanOptimizer` implementation classes to `IcebergPlanOptimizerProvider.getLogicalPlanOptimizers`, which is a return type of `Set<ConnectorPlanOptimizer>`, so I used `Set<ConnectorPlanOptimizer>` to store `planOptimizers` in this piece.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java (hunk 1 above)
Comment (397090770) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergModule.java:171]: Both `IcebergParquetDereferencePushDown` and `IcebergPlanOptimizer` are implementation classes of `ConnectorPlanOptimizer`, so we should also bound `IcebergParquetDereferencePushDown` here.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java (hunk 2 above)
Comment (397090770) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergPageSourceProvider.java:353]: This piece of code actually converts the `ColumnType` of `IcebergColumnHandle` to the `ColumnType` of `HiveColumnHandle`.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizerProvider.java (hunk 3 above)
Comment (397090770) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/optimizer/IcebergPlanOptimizerProvider.java:28]: As explained earlier, we need to pass the `IcebergParquetDereferencePushDown` and `IcebergPlanOptimizer` to the `IcebergPlanOptimizerProvider` class, and finally use it in the `IcebergPlanOptimizerProvider.getLogicalPlanOptimizers() `method.

File: presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergLogicalPlanner.java (hunk 4 above)
Comment (397090770) [reply presto-iceberg/src/test/java/com/facebook/presto/iceberg/TestIcebergLogicalPlanner.java:1]: Yes, I did copy it from `TestHiveLogicalPlanner`

**Comment by 397090770** [comment]
@ChunxuTang Thank you for your review. I have explained some of the above questions, or made some changes. If you have time, please help me to review again. Thanks again.

**Review by wanglinsong — APPROVED** [review]

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergColumnHandle.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergColumnHandle.java:54]: Gotcha. I checked the similar implementation in the `HiveColumnHandle` and let's keep your current implementation.

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChunxuTang) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConnector.java:66]: Thanks for the clarification!

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java [hunk 7]
```diff
@@ -41,6 +41,7 @@
     private int maxPartitionsPerWriter = 100;
     private List<String> hadoopConfigResources = ImmutableList.of();
     private double minimumAssignedSplitWeight = 0.05;
+    private boolean parquetDereferencePushdownEnabled = false;
```
Comment (ChunxuTang) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:44]: Thanks for the update! Could you change it to `true`? Otherwise, there'll be a checkstyle error.

**Review by ChunxuTang — COMMENTED** [review]
@wypb Thanks for the contribution. Generally looks good to me. Just left some very minor suggestions.

@beinan @zhenxiao Would you also like to have a review of this PR?

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java (hunk 7 above)
Comment (wypb) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:44]: I changed it to `private boolean parquetDereferencePushdownEnabled`, in this way checkstyle should be fine

**Review by ChunxuTang — APPROVED** [review]
LGTM. Thanks for the nice work!

**Comment by wypb** [comment]
@ChunxuTang Thank you for your review, I want to know what else do I need to do to merge this MR.

**Review by beinan — APPROVED** [review]
lgtm

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java [hunk 8]
```diff
@@ -152,4 +153,17 @@ public double getMinimumAssignedSplitWeight()
...
+    @ConfigDescription("enable parquet dereference pushdown")
+    public IcebergConfig setParquetDereferencePushdownEnabled(boolean parquetDereferencePushdownEnabled)
+    {
+        this.parquetDereferencePushdownEnabled = parquetDereferencePushdownEnabled;
+        return this;
+    }
+
+    public boolean isParquetDereferencePushdownEnabled()
+    {
+        return this.parquetDereferencePushdownEnabled;
```
Comment (beinan) [inline presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java:167]:
```suggestion
        return parquetDereferencePushdownEnabled;
```

**Review by beinan — APPROVED** [review]
just one minor issue

File: presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java (hunk 8 above)
Comment (wypb) [reply presto-iceberg/src/main/java/com/facebook/presto/iceberg/IcebergConfig.java]: @beinan done, Thank you for your review.

**Comment by beinan** [comment]
rerun the failed test

**Comment by ChunxuTang** [comment]
@wypb @beinan The test failure (native module) looks unrelated to this PR. I noticed that there're some ongoing tasks trying to fix the native test. Probably we'll be able to run all tests successfully after those PRs are merged.

**Comment by wypb** [comment]
HI @ChunxuTang @beinan I triggered the test several times, but the following two tests always fail.
企业微信截图_16844602715050
image
It seems that the folly library cannot be found. Do I need to merge the code on the master branch into my branch?

**Comment by wypb** [comment]
HI @zhenxiao zhenxiao, I resubmitted a PR (https://github.com/prestodb/presto/pull/19710) based on the latest master branch, and it seems that all the tests have passed, please help me to take a look

**Comment by wypb** [comment]
Re-running the tests.
